homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,338 @@
+            await hass.config_entries.async_forward_entry_setup(
+                entry, component)
+
+        # If first connect fails, the next re-connect will be scheduled
+        # outside of _pending_task, in order not to delay HA startup
+        # indefinitely
+        await try_connect(is_disconnect=False)
+
+    hass.async_add_job(complete_setup())
+    return True
+
+
+async def _setup_auto_reconnect_logic(hass: HomeAssistantType, cli, key,
+                                      host):
+    """Set up the re-connect logic for the API client."""
+    from aioesphomeapi.client import APIConnectionError
+
+    async def try_connect(tries=0, is_disconnect=True):
+        """Try connecting to the API client. Will retry if not successful."""
+        if key not in hass.data[DOMAIN]:
+            # When removing/disconnecting manually
+            return
+
+        data = get_entry_data(hass, key)
+        data[DATA_AVAILABLE] = False
+        async_dispatcher_send(hass, DISPATCHER_ON_DEVICE_UPDATE.format(key))
+
+        if tries != 0:
+            # If not first re-try, wait and print message
+            wait_time = min(2**tries, 300)
+            _LOGGER.info("Trying to reconnect in %s seconds.", wait_time)
+            await asyncio.sleep(wait_time)
+
+        if is_disconnect and tries == 0:
+            # This can happen often depending on WiFi signal strength.
+            # So therefore all these connection warnings are logged
+            # as infos. The "unavailable" logic will still trigger so the
+            # user knows if the device is not connected.
+            _LOGGER.info("Disconnected from API.")
+
+        try:
+            await cli.connect()
+            await cli.login()
+        except APIConnectionError as error:
+            _LOGGER.info("Can't connect to esphomelib API for '%s' (%s).",
+                         host, error)
+            # Schedule re-connect in event loop in order not to delay HA
+            # startup. First connect is scheduled in tracked tasks.
+            data['reconnect'] = hass.loop.create_task(
+                try_connect(tries + 1, is_disconnect))
+        else:
+            _LOGGER.info("Successfully connected to %s", host)
+
+    cli.on_disconnect = try_connect
+    return try_connect
+
+
+async def _cleanup_instance(hass: HomeAssistantType, entry: ConfigEntry):
+    """Cleanup the esphomelib client if it exists."""
+    data = hass.data[DOMAIN].pop(entry.entry_id)
+    client = data[DATA_CLIENT]
+    if DATA_RECONNECT_TASK in data:
+        data[DATA_RECONNECT_TASK].cancel()
+    await client.stop()
+
+
+async def async_unload_entry(hass, config_entry) -> bool:
+    """Unload an esphomelib config entry."""
+    await _cleanup_instance(hass, config_entry)
+
+    for component in HA_COMPONENTS:
+        await hass.config_entries.async_forward_entry_unload(
+            config_entry, component)
+
+    return True
+
+
+async def async_setup(hass, config):
```

Comment:
Please move this above `async_setup_entry`. That's where we usually have it, and it also in line with the order that functions will be called.